aggregate.go

```diff
@@ -0,0 +1,440 @@
+		return err
+	}
+
+	// Consider eviction of old data.
+	if time.Since(as.lastEvicted) > as.MinEvictAge {
+		if err := as.evictBefore(time.Now().Add(-1 * as.MinEvictAge)); err != nil {
+			return err
+		}
+	}
+
+	// Grab the group for our span.
+	group, ok := as.group(id, anns...)
+	if !ok {
+		// We don't have a group for the trace, and can't create one (the
+		// spanName event isn't present yet).
+		return nil
+	}
+
+	// Unmarshal the events.
+	var events []Event
+	if err := UnmarshalEvents(anns, &events); err != nil {
+		return err
+	}
+
+	// Find the start and end time of the trace.
+	var (
+		eStart, eEnd time.Time
+		haveTimes    = false
+	)
+	for _, e := range events {
+		e, ok := e.(TimespanEvent)
+		if !ok {
+			continue
+		}
+		if !haveTimes {
+			haveTimes = true
+			eStart = e.Start()
+			eEnd = e.End()
+			continue
+		}
+		if v := e.Start(); v.UnixNano() < eStart.UnixNano() {
+			eStart = v
+		}
+		if v := e.End(); v.UnixNano() > eEnd.UnixNano() {
+			eEnd = v
+		}
+	}
+	if !haveTimes {
+		// We didn't find any timespan events at all, so we're done here.
+		return nil
+	}
+
+	// Update the group to consider this trace being one of the slowest.
+	group.update(eStart, eEnd, id.Trace, func(trace ID) {
+		// Delete the request trace from the output store.
+		if err := as.MemoryStore.Delete(trace); err != nil {
+			log.Printf("AggregateStore: failed to delete a trace: %s", err)
+		}
+	})
+
+	// Move traces from the limit store into the group, as needed.
+	for _, slowest := range group.Slowest {
+		// Find the trace in the limit store.
+		trace, err := as.pre.Trace(slowest.TraceID)
+		if err == ErrTraceNotFound {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+
+		// Place into output store.
+		var walk func(t *Trace) error
+		walk = func(t *Trace) error {
+			err := as.MemoryStore.Collect(t.Span.ID, t.Span.Annotations...)
+			if err != nil {
+				return err
+			}
+			for _, sub := range t.Sub {
+				if err := walk(sub); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		if err := walk(trace); err != nil {
+			return err
+		}
+
+		// Delete from the limit store.
+		err = as.pre.Delete(slowest.TraceID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// As we're updating the aggregation event, we go ahead and delete the old
+	// one now. We do this all under as.MemoryStore.Lock otherwise users (e.g. the
+	// web UI) can pull from as.MemoryStore when the trace has been deleted.
+	as.MemoryStore.Lock()
+	if err := as.MemoryStore.deleteNoLock(group.Trace); err != nil {
+		return err
+	}
+
+	// Record an aggregate event with the given name.
+	recEvent := func(e Event) error {
+		anns, err := MarshalEvent(e)
+		if err != nil {
+			return err
+		}
+		return as.MemoryStore.collectNoLock(SpanID{Trace: group.Trace}, anns...)
+	}
+	if err := recEvent(spanName{Name: group.Name}); err != nil {
+		return err
+	}
+	ev := &AggregateEvent{
+		Name:  group.Name,
+		Times: group.Times,
+	}
+	for _, slowest := range group.Slowest {
```

Comment:
Is it possible to precompute this outside of the lock. You want the time in the lock to be as tight as possible.